File: is_substring/is_substring.js


let isSubsequence = (s,t) => {
  if(s.length > t.length){
    return false;
  }

  let s_index = 0
  let t_index = 0 

  while (s_index < s.length && t_index < t.length) {
    if (s[s_index] === t[t_index]){
      s_index ++;
    }
    t_index++;
  }
  return s_index === s.length;
};

let s = "sing"
let t = "string"

let a = "fly"
let b = "flow"

console.log(isSubsequence(s, t));
console.log(isSubsequence(a, b));

// const string = "string"

// console.log(string.indexOf("s", 2))
// console.log(string[string.length-1])

const grades = [{grade: 10}, {grade: 15}, {grade: 5}];
console.log(grades.reduce((total, current) => {
  console.log(current);
  return total + current.grade // visualize that this is an object
}, 0));

File: is_substring/is_substring.rb
string_1 = "string"
string_2 = "sing"
string_3 = "ings"
bad_string = "sang"

def is_subsequence(s, t)
  # don't need array's since we're just checking the length of the string
  # you could check first that s is not longer than t in which case it couldn't be a subsequence
  # something like
  #return false if s.length > t.length
  # s_arr = s.chars
  # t_arr = t.chars

  s_index = 0
  t_index = 0

  while s_index < s_arr.length && t_index < t_arr.length
    if t[t_index] == s[s_index]
      s_index += 1
    end
    t_index += 1
  end
    
  return s_index == s_arr.length

end

puts is_subsequence(string_2, string_1)